src/runtime/virtcontainers/qemu_amd64.go

```diff
@@ -191,3 +206,54 @@ func (q *qemuAmd64) appendImage(ctx context.Context, devices []govmmQemu.Device,
 func (q *qemuAmd64) appendBridges(devices []govmmQemu.Device) []govmmQemu.Device {
 	return genericAppendBridges(devices, q.Bridges, q.qemuMachine.Type)
 }
+
+// enable protection
+func (q *qemuAmd64) enableProtection() error {
+	flags, err := CPUFlags(procCPUInfo)
+	if err != nil {
+		return err
+	}
+
+	// TDX is supported and properly loaded when the firmware directory exists or `tdx` is part of the CPU flags
+	if d, err := os.Stat(tdxSysFirmwareDir); (err == nil && d.IsDir()) || flags[tdxCPUFlag] {
```

Comment:
Could we encapsulate the protection technology "detection" in an hypervisor.go global function:

```Go
func availableGuestProtection(cpuFlags map[string]bool) (guestProtection, error)
```

And then the code below would look like:

```Go

q.protection, err := availableGuestProtection(flags)
if err != nil {
    return err
}

switch q.protection {
   case tdxProtection:
              if q.qemuMachine.Options != "" {
			q.qemuMachine.Options += ","
		}
		q.qemuMachine.Options += "kvm-type=tdx,confidential-guest-support=tdx"
		q.kernelParams = append(q.kernelParams, Param{"tdx_guest", ""})
		virtLog.WithFields(logrus.Fields{
			"subsystem":     "qemuAmd64",
			"machine":       q.qemuMachine,
			"kernel-params": q.kernelParameters}).
			Info("Enabling TDX guest protection")
		return nil
    default:
       return fmt.Errorf("This system doesn't support Confidential Computing (Guest Protection)")
}
```

Reply:
technology detection is architecture specific, I wonder if using only cpu flags is enough to detect the technology in **all** the architectures

@jimcadden @Jakob-Naucke @bpradipt  what do you think? does it work for you or you prefer to keep it technology detection as something arch specific (`qemu_${ARCH}.go`) ?

Reply:
@devimc, for SEV we'll want to confirm both that it's supported by the hardware and it's enabled by the host kernel. Cpu flags tells us the first bit. We can get both from the host filesystem, which I believe we can do from within the existing switch statement 
> $ cat /sys/module/kvm_amd/parameters/sev 
> 1

Reply:
@devimc On s390x, we need to check for two things.
1. s390x uses _CPU facility bits_ (not to be confused with flags), one of which must be active to signal Secure Execution support. These facilities can e.g. be retrieved from `/proc/cpuinfo`. For the time being, I had written a function `CPUFacilities` in (the newly created) [`hypervisor_s390x.go`](https://github.com/kata-containers/kata-containers/pull/1772/files#diff-b9ae770572724ed09cc6a2fb15a210aa64080affcead76fea6a5195a4a98e27a).
2. `prot_virt=1` must be on the kernel command line. For checking this, I have added a [function](https://github.com/kata-containers/kata-containers/pull/1772/files#diff-7c48f6b46053aca2fcf0873f97a38587d9477cfa2aa404daec37b87feb02ac2dR729) in `hypervisor.go`.

I agree that it would be cleaner to have this detection mechanism in `hypervisor` scope instead of `qemu` scope. 
However, without an `ArchBase` interface like we have for QEMU, we could only call common code functions, but `CPUFacilities` doesn't really belong into common code.
Alternatively, we could put a check function in each of `hypervisor_{arch}.go` (but none of those exist right now, at least not upstream).

Reply:
@Jakob-Naucke @jimcadden @sameo thanks, change applied, ptal

Reply:
@devimc I need to check for something that is only really valid on s390x (see (1)), so with the suggested change, how would I do that? Also, `CPUFlags` is going to error on s390x, because `/proc/cpuinfo` has no `flags` field on s390x.

Reply:
@Jakob-Naucke okay, `availableGuestProtection` is arch specific and we'll need `hypervisor_${ARCH}.go` files for this function (and maybe others), what do you think?

cc @sameo

Reply:
@devimc I think that is the cleanest since `availableGuestProtection` won't have much in common across architectures anyways.

Reply:
@Jakob-Naucke done, ptal

Reply:
Thank you @devimc. This works for me (see also updated #1772).
@sameo This is technically not quite what you asked for, since it is not common `hypervisor.go` functionality, but branched out to architectures. However, with the different ways architectures handle CPU info, I think this is the only way we can go. Does that sound reasonable? I still prefer having it in `hypervisor` instead of `qemu`.